### ISSUE TITLE
[hl/C] Comparison of dynamically typed floats and hl.F32 fails

### ISSUE THREAD

**Issue by MoritzBrueckner** [HaxeFoundation/haxe#9869]
There is a bug in comparing two floats of different types when they are stored in variables of the `Dynamic` type.

**Minimal example code**:
```haxe
var a: kha.FastFloat = 4.123; // From the Kha framework. On hl/C target, this is a typedef to hl.F32
var b: Float = 4.120;

// Without those two lines, both static and non-static targets would print 'true', which is the wanted behaviour
var aD: Dynamic = a;
var bD: Dynamic = b;

// Non-static target: true, hl/C: false. true should be printed on all targets
trace(aD >= bD);
```

The same bug might occur with other 32bit floats on other targets as well, I don't know.

Also, casting FastFloat/hl.F32 values that are stored in dynamic variables back to floats does not always work on the hl target, but I wasn't able to find a minimal reproducible example yet. The error message is `Unsupported type value Single` when compiling .hx code like `v1 = cast(v1, Float);` (`v1` is a dynamic variable that holds a FastFloat).

More background information for this issue can be found here: https://github.com/Kode/Kha/issues/1237, https://github.com/armory3d/armory/issues/1814

**Edit:** Might be related to https://github.com/HaxeFoundation/hashlink/issues/411 and https://github.com/HaxeFoundation/hashlink/issues/396

Thank you!

**Comment by ghost** [comment]
Haxe `Float` type in Hashlink is double or `hl.F64` not `hl.F32`.

**Comment by MoritzBrueckner** [comment]
Yet the comparison works without the `Dynamic` type. I'm not well informed about the bitwise differences between 32 and 64 bit floats, but if they aren't comparable naturally, shouldn't there be an implicit cast?

**Comment by ghost** [comment]
The reason behind why `trace(aD >= bD);` prints false is because of haxe implementation of abstract types and their conversions, when you make `Dynamic` type, it's Dynamic and does not have implicit value. Comparison between types is that type job to do.
Haxe abstract types define how that type compare itself with another type but Dynamic does not do this kind of [operator overloading](https://haxe.org/manual/types-abstract-operator-overloading.html).

And to answer your question on casting, it's working fine with `hl.F32` and `Float`. If it's not working correctly it's Kha issue so you have to open an issue in Kha repo.

This works fine:
```haxe
var aDc:hl.F32 = cast(aD, hl.F32);
var bDc:Float = cast(bD, Float);
trace(aDc >= bDc); // true
```

**Comment by MoritzBrueckner** [comment]
Ok, thanks for clarification!

> And to answer your question on casting, it's working fine with hl.F32 and Float. If it's not working correctly it's Kha issue so you have to open an issue in Kha repo.

I did open an issue there which was closed and I was referred to the Haxe repo (FastFloat is really just a typedef). But as I said I'm not able to recreate that casting error in a minimal example so I guess I just have to find a workaround and hope someday some other person stumbles on the same issue and finds a way to reproduce that particular issue so that it can be fixed.

Thank you!
